runtime/src/main/kotlin/kotlin/native/concurrent/WorkerBoundReference.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+
+@SymbolName("Kotlin_WorkerBoundReference_create")
+external private fun createWorkerBoundReference(value: Any): NativePtr
+
+@SymbolName("Kotlin_WorkerBoundReference_deref")
+external private fun derefWorkerBoundReference(ref: NativePtr): Any
+
+/**
+ * A frozen shared reference to a Kotlin object
+ *
+ * Can be safely passed between workers, but can only be accessed on the worker it was created on,
+ * unless the referred object is frozen too
+ *
+ * Note: Garbage collector currently cannot free any reference cycles with [WorkerBoundReference] or [DisposableWorkerBoundReference] in them.
+ * Consider using [DisposableWorkerBoundReference] to manually resolve cycles with explicit calls to [DisposableWorkerBoundReference.dispose]
+ */
+@Frozen
+@NoReorderFields
+@ExportTypeInfo("theWorkerBoundReferenceTypeInfo")
+public class WorkerBoundReference<out T : Any>(value: T) {
+
+    private val ptr = createWorkerBoundReference(value)
+
+    /**
+     * The referenced value.
+     * @throws IncorrectDereferenceException if referred object is not frozen and this is called from a different worker, than [WorkerBoundReference] was created on
+     */
+    val value: T
```

Comment:
Do we need something like `valueOrNull` which doesn't throw in other workers, but returns null instead?
cc @elizarov

Reply:
You can take a look at my use-cases here: https://github.com/Kotlin/kotlinx.coroutines/blob/native-mt/kotlinx-coroutines-core/native/src/internal/Sharing.kt#L199

As you can see, the most important additional thing I need is a reference to the `worker`.

Reply:
However, `valueOrNull` would be occasionally useful, too. See usages of this function, for example: https://github.com/Kotlin/kotlinx.coroutines/blob/native-mt/kotlinx-coroutines-core/native/src/internal/Sharing.kt#L159

Reply:
Implemented in https://github.com/JetBrains/kotlin-native/pull/4115